From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kevin Venalainen <kevin.venalainen@around.co>
Date: Fri, 26 Feb 2021 16:57:19 -0800
Subject: adds around runner test executable

    This patch adds a test executable to allow running WebRTC pipeline offline

Signed-off-by: gellert <gellert.hegyi@around.co>

diff --git a/modules/audio_processing/BUILD.gn b/modules/audio_processing/BUILD.gn
index d57d29589b790ed91f0480f24c09b1bccb8601aa..3935d5e376e3ee34fbea9da79c4a2ce6c772467f 100644
--- a/modules/audio_processing/BUILD.gn
+++ b/modules/audio_processing/BUILD.gn
@@ -608,3 +608,40 @@ rtc_library("audioproc_test_utils") {
   ]
   absl_deps = [ "//third_party/abseil-cpp/absl/types:optional" ]
 }
+
+rtc_executable("around_runner") {
+  testonly = true
+  visibility = ["*"]
+
+  configs += [
+    ":apm_debug_dump",
+  ]
+
+  deps = [
+    ":api",
+    ":audio_buffer",
+    ":audio_frame_view",
+    ":audio_processing",
+    ":audioproc_test_utils",
+    "../../api:array_view",
+    "../../api:scoped_refptr",
+    "../../api/audio:aec3_config",
+    "../../api/audio:aec3_factory",
+    "../../common_audio",
+    "../../common_audio:common_audio_c",
+    "../../system_wrappers",
+    "../../system_wrappers:field_trial",
+    "../../system_wrappers:metrics",
+    "../../pc:peerconnection",
+    "around_runner:around_runner_src",
+  ]
+
+  defines = [
+  ]
+
+  if (rtc_prefer_fixed_point) {
+    defines += [ "WEBRTC_AUDIOPROC_FIXED_PROFILE" ]
+  } else {
+    defines += [ "WEBRTC_AUDIOPROC_FLOAT_PROFILE" ]
+  }
+}
diff --git a/modules/audio_processing/around_runner/BUILD.gn b/modules/audio_processing/around_runner/BUILD.gn
new file mode 100644
index 0000000000000000000000000000000000000000..c1d77f5fc2c4aa22ca325231ebc65431925863a8
--- /dev/null
+++ b/modules/audio_processing/around_runner/BUILD.gn
@@ -0,0 +1,13 @@
+import("../../../webrtc.gni")
+
+rtc_source_set("around_runner_src") {
+  visibility = [ "../*" ]
+  sources = [
+    "mocks.h",
+    "around_runner.cc",
+    "ketopt.h",
+    "around_args.h",
+  ]
+  deps = []
+  absl_deps = [ "//third_party/abseil-cpp/absl/types:optional" ]
+}
diff --git a/modules/audio_processing/around_runner/around_args.h b/modules/audio_processing/around_runner/around_args.h
new file mode 100644
index 0000000000000000000000000000000000000000..0295086aff9e00031c8fac9477b78ef3498aaba0
--- /dev/null
+++ b/modules/audio_processing/around_runner/around_args.h
@@ -0,0 +1,127 @@
+#pragma once
+
+#include <stdio.h>
+#include <string>
+#include <iostream>
+#include "ketopt.h"
+#include "absl/types/optional.h"
+
+using namespace std;
+
+#if !defined(endline)
+# if defined(WEBRTC_WIN)
+#   define endline "\r\n"
+# else
+#   define endline "\n"
+# endif
+#endif
+
+struct around_args {
+  bool fullyParsed = true;
+  string input = "";
+  string loopback = "";
+  string output = "";
+  bool verbose = false;
+
+  absl::optional<bool> aec; //WebRTC AEC
+  absl::optional<bool> ns; //WebRTC NS
+  absl::optional<int> ns_level; // WebRTC NS::level
+  absl::optional<float> rnnoise_level; // Around RNNoise maximum suppression limit
+
+  string ToString() const {
+    string s;
+    s.reserve(256);
+    s
+      .append(" input: ").append(input).append(endline)
+      .append(" loopback: ").append(loopback).append(endline)
+      .append(" output: ").append(output).append(endline)
+      .append(" verbose: ").append(to_string(verbose)).append(endline)
+      .append(" aec: ").append(aec ? to_string(*aec) : "(default)").append(endline)
+      .append(" ns: ").append(ns ? to_string(*ns) : "(default)").append(endline)
+      .append(" ns_level: ").append(ns_level ? to_string(*ns_level) : "(default)").append(endline)
+      .append(" rnnoise_level: ").append(rnnoise_level ? to_string(*rnnoise_level) : "(default)").append(endline);
+    return s;
+  }
+};
+
+std::string HelpString()
+{
+  string s;
+  s.reserve(1024);
+  s
+    .append("Around Runner help page:").append(endline)
+    .append("-i,--input: Input .wav file, represents microphone signal").append(endline)
+    .append("-l,--loopback: Loopback speaker .wav file, unused if AEC is disabled").append(endline)
+    .append("-o,--output: Output .wav file").append(endline)
+    .append("-v,--verbose: Flag setting verbose output if present").append(endline)
+    .append("--aec=x: Forces feature to disabled if x=0, else enabled. Feature default if not present.").append(endline)
+    .append("--ns=x: Forces feature to disabled if x=0, else enabled. Feature default if not present.").append(endline)
+    .append("--ns_level=x: Sets NoiseSuppression::Level, with argument corresponding to the enum values. (Low = 0, VeryHigh = 3)").append(endline)
+    .append("--rnnoise_level=x: Sets maximum RNNoise suppression, with argument corresponding to dB. (Default = 15.0, Off = 0.0, Low = 10.0)").append(endline)
+  ;
+  return s;
+}
+
+enum longargs : int {
+  input = 301, // Ensure greater than the largest ASCII character code
+  loopback,
+  output,
+  verbose,
+  aec,
+  ns,
+  ns_level,
+  rnnoise_level,
+};
+
+#define LONGOPT(x, arg_type) {#x, arg_type, x}
+
+static const ko_longopt_t longopts[] = {
+  LONGOPT(input, ko_required_argument),
+  LONGOPT(loopback, ko_required_argument),
+  LONGOPT(output, ko_required_argument),
+  LONGOPT(verbose, ko_no_argument),
+  LONGOPT(aec, ko_required_argument),
+  LONGOPT(ns, ko_required_argument),
+  LONGOPT(ns_level, ko_required_argument),
+  LONGOPT(rnnoise_level, ko_required_argument),
+};
+
+static const char shortopts[] = "i:l:o:v";
+
+around_args parse_args(int argc, char* argv[])
+{
+  around_args args;
+  ketopt_t opt = KETOPT_INIT;
+  int c;
+  while ((c = ketopt(&opt, argc, argv, true, shortopts, longopts)) >= 0) {
+    switch (c)
+    {
+    case 'i':
+    case longargs::input:
+      args.input = opt.arg; break;
+    case 'l':
+    case longargs::loopback:
+      args.loopback = opt.arg; break;
+    case 'o':
+    case longargs::output:
+      args.output = opt.arg; break;
+    case 'v':
+    case longargs::verbose:
+      args.verbose = true; break;
+    case longargs::aec:
+      args.aec = atoi(opt.arg); break;
+    case longargs::ns:
+      args.ns = atoi(opt.arg); break;
+    case longargs::ns_level:
+      args.ns_level = atoi(opt.arg); break;
+    case longargs::rnnoise_level:
+      args.rnnoise_level = static_cast<float>(atof(opt.arg)); break;
+    case '?':
+      cerr << "Unknown option: " << (opt.opt ? (char)opt.opt : ':') << endl; args.fullyParsed = false; break;
+    case ':':
+      cerr << "Missing argument: " << (opt.opt ? (char)opt.opt : ':') << endl; args.fullyParsed = false; break;
+    }
+  }
+  return args;
+}
+
diff --git a/modules/audio_processing/around_runner/around_runner.cc b/modules/audio_processing/around_runner/around_runner.cc
new file mode 100644
index 0000000000000000000000000000000000000000..fd6f500be940cabf3a3152857f888b33009be13f
--- /dev/null
+++ b/modules/audio_processing/around_runner/around_runner.cc
@@ -0,0 +1,330 @@
+#include <cstddef>
+#include <iostream>
+
+#ifndef WEBRTC_APM_DEBUG_DUMP
+#define WEBRTC_APM_DEBUG_DUMP 1
+#endif
+
+#define UNREFERENCED_PARAMETER(x) (x)
+
+#define IF_TRUE_LOGERR_RETURN(condition)                        \
+    if (condition)                                              \
+    {                                                           \
+        cerr << "Error: " << #condition << endline;                \
+        return -1;                                              \
+    }
+
+#if !defined(endline)
+# if defined(WEBRTC_WIN)
+#   define endline "\r\n"
+# else
+#   define endline "\n"
+# endif
+#endif
+
+#include "mocks.h"
+#include "../include/audio_processing.h"
+#include "../audio_processing_impl.h"
+#include "rtc_base/logging.h"
+#include "rtc_base/ref_counted_object.h"
+#include "common_audio/wav_file.h"
+#include "../test/audio_processing_builder_for_testing.h"
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wunused-function"
+#pragma GCC diagnostic ignored "-Wheader-hygiene"
+#include "around_args.h"
+#pragma GCC diagnostic pop
+
+using namespace webrtc;
+using namespace rtc;
+using namespace std;
+
+class AudioProcessingImplComplete : public AudioProcessingImpl {
+ public:
+  explicit AudioProcessingImplComplete(const webrtc::Config& config)
+    : AudioProcessingImpl(config) {}
+
+  void AddRef() const override
+  {
+  }
+
+  RefCountReleaseStatus Release() const override
+  {
+    return RefCountReleaseStatus::kOtherRefsRemained;
+  }
+};
+
+unique_ptr<WavReader> MakeWavReader(const string& filename)
+{
+  cout << "opening file: " << filename << endline;
+  auto res = make_unique<WavReader>(filename);
+  cout << "Format: " << res->num_channels() << "ch/" << res->sample_rate() << "hz" << endline;
+  return res;
+}
+
+unique_ptr<WavWriter> MakeWavWriter(const string& filename,
+                                    int sample_rate,
+                                    size_t num_channels)
+{
+  cout << "writing file: " << filename << endline
+    << "Format: " << num_channels << "ch/" << sample_rate << "hz" << endline;
+  return make_unique<WavWriter>(filename, sample_rate, num_channels, WavFile::SampleFormat::kFloat);
+}
+
+// Deinterleaves sample-interleaved data to channel-sequential blocks. Pointers must not overlap.
+void Deinterleave(const float* in, float* out, const int frames, const int channels)
+{
+  // Helper pointers to output per channel.
+  vector<float*> pOut(channels);
+  for (int i = 0; i < channels; i++)
+  {
+    pOut[i] = out + (i * frames);
+  }
+
+  for (int f = 0; f < frames; f++)
+  {
+    for (int c = 0; c < channels; c++)
+    {
+      *pOut[c]++ = *in++;
+    }
+  }
+}
+
+// Interleaves channel-sequential data to sample-interleaved data. Pointers must not overlap.
+void Interleave(const float* in, float* out, const int frames, const int channels)
+{
+  // Helper pointers to input per channel.
+  vector<const float*> pIn(channels);
+  for (int i = 0; i < channels; i++)
+  {
+    pIn[i] = in + (i * frames);
+  }
+
+  for (int f = 0; f < frames; f++)
+  {
+    for (int c = 0; c < channels; c++)
+    {
+      *out++ = *pIn[c]++;
+    }
+  }
+}
+
+// Scales input data by a value, writing to out. In and out are OK to overlap.
+void Scale(const float* in, float* out, const float scaleVal, const int frames, const int channels)
+{
+  for (int i = 0; i < frames * channels; i++)
+  {
+    out[i] = in[i] * scaleVal;
+  }
+}
+
+AudioProcessing::Config& SetConfigDefaults(AudioProcessing::Config& config)
+{
+  config.pipeline.maximum_internal_processing_rate = 48000;
+  config.pipeline.multi_channel_render = true;
+  config.pipeline.multi_channel_capture = false;
+
+  config.pre_amplifier.enabled = false;
+
+  config.high_pass_filter.enabled = true;
+
+  config.echo_canceller.enabled = true;
+
+  config.noise_suppression.enabled = true;
+  config.noise_suppression.level = AudioProcessing::Config::NoiseSuppression::kModerate;
+
+  config.transient_suppression.enabled = true;
+
+  config.voice_detection.enabled = false;
+
+  config.gain_controller1.enabled = true;
+  config.gain_controller1.mode = AudioProcessing::Config::GainController1::kAdaptiveDigital;
+  config.gain_controller1.analog_gain_controller.enabled = false;
+
+  config.gain_controller2.enabled = false;
+
+  config.residual_echo_detector.enabled = false;
+
+  config.level_estimation.enabled = false;
+
+  config.around_audio_processing.enabled = true;
+
+  return config;
+}
+
+int main(int argc, char* argv[])
+{
+  if (argc <= 1)
+  {
+    cout << HelpString() << endline;
+    return 0;
+  }
+
+  auto args = parse_args(argc, argv);
+
+  if (!args.fullyParsed)
+  {
+    cout << "Failed to parse some args, aborting. Check that long options have '--' instead of '-'." << endline;
+    cout << "Parsed arguments:" << endline << args.ToString() << endline;
+    return -1;
+  }
+
+  if (args.verbose)
+  {
+    cout << args.ToString() << endline;
+    rtc::LogMessage::LogToDebug(rtc::LS_VERBOSE);
+  }
+
+  // If no loopback file provided, pretend silence.
+  bool useLoopbackFile = args.loopback != "" ? true : false;
+
+  // Input files
+  cout << "[Input] ";
+  auto inFile = MakeWavReader(args.input);
+  std::unique_ptr<WavReader> revInFile;
+  if (useLoopbackFile)
+  {
+    cout << "[Loopback] ";
+    revInFile = MakeWavReader(args.loopback);
+  }
+  else
+  {
+    cout << "[Loopback] No file provided, using silence" << endline;
+  }
+
+  // Format helpers
+  auto input_sample_rate = inFile->sample_rate();
+  auto input_channels = inFile->num_channels();
+  auto output_sample_rate = input_sample_rate;
+  auto output_channels = input_channels;
+  auto rev_input_sample_rate = useLoopbackFile ? revInFile->sample_rate() : input_sample_rate;
+  auto rev_input_channels = useLoopbackFile ? revInFile->num_channels() : input_channels;
+  auto rev_output_sample_rate = rev_input_sample_rate;
+  auto rev_output_channels = rev_input_channels;
+
+  // Make the APM
+  rtc::scoped_refptr<AudioProcessingImpl> ap =
+      new rtc::RefCountedObject<AudioProcessingImpl>(webrtc::Config());
+  IF_TRUE_LOGERR_RETURN(AudioProcessing::kNoError != ap->Initialize());
+
+  // Set stream formats
+  ProcessingConfig proc;
+  proc.input_stream() = StreamConfig(input_sample_rate, input_channels, false);
+  proc.output_stream() = StreamConfig(output_sample_rate, output_channels, false);
+  proc.reverse_input_stream() = StreamConfig(rev_input_sample_rate, rev_input_channels, false);
+  proc.reverse_output_stream() = StreamConfig(rev_output_sample_rate, rev_output_channels, false);
+  ap->Initialize(proc);
+
+  // Apply any needed pipeline config
+  auto config = ap->GetConfig();
+  SetConfigDefaults(config);
+  if (args.aec) config.echo_canceller.enabled = *args.aec;
+  if (args.ns) config.noise_suppression.enabled = *args.ns;
+  if (args.ns_level) config.noise_suppression.level =
+    static_cast<AudioProcessing::Config::NoiseSuppression::Level>(*args.ns_level);
+  if (args.rnnoise_level) config.around_audio_processing.noise_suppression_db = *args.rnnoise_level;
+  ap->ApplyConfig(config);
+
+  // Output files
+  cout << "[Output]: ";
+  auto outFile = MakeWavWriter(args.output, proc.output_stream().sample_rate_hz(), proc.output_stream().num_channels());
+
+  // Print config
+  if (args.verbose)
+  {
+    cout << ap->GetConfig().ToString() << endline;
+  }
+
+  // Setup needed buffers
+  vector<float> inInterleaved(proc.input_stream().num_samples(), 0.0f);
+  vector<float> outInterleaved(proc.output_stream().num_samples(), 0.0f);
+  vector<float> revInInterleaved(proc.reverse_input_stream().num_samples(), 0.0f);
+  vector<float> revOutInterleaved(proc.reverse_output_stream().num_samples(), 0.0f);
+
+  vector<float> inDeinterleaved(proc.input_stream().num_samples(), 0.0f);
+  vector<float> outDeinterleaved(proc.output_stream().num_samples(), 0.0f);
+  vector<float> revInDeinterleaved(proc.reverse_input_stream().num_samples(), 0.0f);
+  vector<float> revOutDeinterleaved(proc.reverse_output_stream().num_samples(), 0.0f);
+
+  vector<float*> inBuffers(proc.input_stream().num_channels());
+  vector<float*> outBuffers(proc.output_stream().num_channels());
+  vector<float*> revInBuffers(proc.reverse_input_stream().num_channels());
+  vector<float*> revOutBuffers(proc.reverse_output_stream().num_channels());
+
+  for (size_t i = 0; i < inBuffers.size(); i++)
+  { inBuffers[i] = &inDeinterleaved[i * proc.input_stream().num_frames()]; }
+  for (size_t i = 0; i < outBuffers.size(); i++)
+  { outBuffers[i] = &outDeinterleaved[i * proc.output_stream().num_frames()]; }
+  for (size_t i = 0; i < revInBuffers.size(); i++)
+  { revInBuffers[i] = &revInDeinterleaved[i * proc.reverse_input_stream().num_frames()]; }
+  for (size_t i = 0; i < revOutBuffers.size(); i++)
+  { revOutBuffers[i] = &revOutDeinterleaved[i * proc.reverse_output_stream().num_frames()]; }
+
+  // Process
+  size_t inSamplesProcessed = 0;
+  size_t revSamplesProcessed = 0;
+  bool moreData = true;
+  while (moreData)
+  {
+    // Read new file data
+    size_t numInputSamplesActual = inFile->ReadSamples(proc.input_stream().num_samples(), inInterleaved.data());
+    if (numInputSamplesActual < proc.input_stream().num_samples())
+      moreData = false;
+    size_t numRevInputSamplesActual = 0U;
+    if (useLoopbackFile)
+    {
+      numRevInputSamplesActual = revInFile->ReadSamples(proc.reverse_input_stream().num_samples(), revInInterleaved.data());
+      if (numRevInputSamplesActual < proc.reverse_input_stream().num_samples())
+        moreData = false;
+    }
+
+    // Deinterleave
+    Deinterleave(inInterleaved.data(), inDeinterleaved.data(),
+      proc.input_stream().num_frames(), proc.input_stream().num_channels());
+    Deinterleave(revInInterleaved.data(), revInDeinterleaved.data(),
+      proc.reverse_input_stream().num_frames(), proc.reverse_input_stream().num_channels());
+
+    // Scale, since WebRTC processing needs [0, 1] and WavReader scales to [0, 32767]
+    Scale(inDeinterleaved.data(), inDeinterleaved.data(), 1.0f/32767.0f,
+      proc.input_stream().num_frames(), proc.input_stream().num_channels());
+    Scale(revInDeinterleaved.data(), revInDeinterleaved.data(), 1.0f/32767.0f,
+      proc.reverse_input_stream().num_frames(), proc.reverse_input_stream().num_channels());
+
+    // Process
+    int processResult = AudioProcessing::kNoError;
+    processResult = ap->ProcessStream(inBuffers.data(), proc.input_stream(), proc.output_stream(), outBuffers.data());
+    if (processResult != AudioProcessing::kNoError)
+    {
+      cout << "ProcessStream error: " << processResult << endline;
+      moreData = false;
+    }
+    processResult = ap->ProcessReverseStream(revInBuffers.data(), proc.reverse_input_stream(), proc.reverse_output_stream(), revOutBuffers.data());
+    if (processResult != AudioProcessing::kNoError)
+    {
+      cout << "ProcessReverseStream error: " << processResult << endline;
+      moreData = false;
+    }
+
+    // Interleave
+    Interleave(outDeinterleaved.data(), outInterleaved.data(),
+      proc.output_stream().num_frames(), proc.output_stream().num_channels());
+    Interleave(revOutDeinterleaved.data(), revOutInterleaved.data(),
+      proc.reverse_output_stream().num_frames(), proc.reverse_output_stream().num_channels());
+
+    // Re-scale back to [0, 32767] since WavWriter expects this range.
+    Scale(outInterleaved.data(), outInterleaved.data(), 32767.0f,
+      proc.output_stream().num_frames(), proc.output_stream().num_channels());
+
+    // Write
+    outFile->WriteSamples(outInterleaved.data(), proc.output_stream().num_samples());
+    // Not bothering to write the reverse output, it's the same as loopback right now.
+
+    inSamplesProcessed += numInputSamplesActual;
+    revSamplesProcessed += numRevInputSamplesActual;
+  }
+
+  cout << "Input frames processed: " << inSamplesProcessed / proc.input_stream().num_channels() << endline
+       << "Reverse frames processed: " << revSamplesProcessed / proc.reverse_input_stream().num_channels() << endline;
+  return 0;
+}
diff --git a/modules/audio_processing/around_runner/ketopt.h b/modules/audio_processing/around_runner/ketopt.h
new file mode 100644
index 0000000000000000000000000000000000000000..b116c1507e2c77e6959b48ce781b939158bb0769
--- /dev/null
+++ b/modules/audio_processing/around_runner/ketopt.h
@@ -0,0 +1,120 @@
+#ifndef KETOPT_H
+#define KETOPT_H
+
+#include <string.h> /* for strchr() and strncmp() */
+
+#define ko_no_argument       0
+#define ko_required_argument 1
+#define ko_optional_argument 2
+
+typedef struct {
+	int ind;   /* equivalent to optind */
+	int opt;   /* equivalent to optopt */
+	const char *arg; /* equivalent to optarg */
+	int longidx; /* index of a long option; or -1 if short */
+	/* private variables not intended for external uses */
+	int i, pos, n_args;
+} ketopt_t;
+
+typedef struct {
+	const char *name;
+	int has_arg;
+	int val;
+} ko_longopt_t;
+
+static ketopt_t KETOPT_INIT = { 1, 0, 0, -1, 1, 0, 0 };
+
+static void ketopt_permute(char *argv[], int j, int n) /* move argv[j] over n elements to the left */
+{
+	int k;
+	char *p = argv[j];
+	for (k = 0; k < n; ++k)
+		argv[j - k] = argv[j - k - 1];
+	argv[j - k] = p;
+}
+
+/**
+ * Parse command-line options and arguments
+ *
+ * This fuction has a similar interface to GNU's getopt_long(). Each call
+ * parses one option and returns the option name.  s->arg points to the option
+ * argument if present. The function returns -1 when all command-line arguments
+ * are parsed. In this case, s->ind is the index of the first non-option
+ * argument.
+ *
+ * @param s         status; shall be initialized to KETOPT_INIT on the first call
+ * @param argc      length of argv[]
+ * @param argv      list of command-line arguments; argv[0] is ignored
+ * @param permute   non-zero to move options ahead of non-option arguments
+ * @param ostr      option string
+ * @param longopts  long options
+ *
+ * @return ASCII for a short option; ko_longopt_t::val for a long option; -1 if
+ *         argv[] is fully processed; '?' for an unknown option or an ambiguous
+ *         long option; ':' if an option argument is missing
+ */
+static int ketopt(ketopt_t *s, int argc, char *argv[], int permute, const char *ostr, const ko_longopt_t *longopts)
+{
+	int opt = -1, i0, j;
+	if (permute) {
+		while (s->i < argc && (argv[s->i][0] != '-' || argv[s->i][1] == '\0'))
+			++s->i, ++s->n_args;
+	}
+	s->arg = 0, s->longidx = -1, i0 = s->i;
+	if (s->i >= argc || argv[s->i][0] != '-' || argv[s->i][1] == '\0') {
+		s->ind = s->i - s->n_args;
+		return -1;
+	}
+	if (argv[s->i][0] == '-' && argv[s->i][1] == '-') { /* "--" or a long option */
+		if (argv[s->i][2] == '\0') { /* a bare "--" */
+			ketopt_permute(argv, s->i, s->n_args);
+			++s->i, s->ind = s->i - s->n_args;
+			return -1;
+		}
+		s->opt = 0, opt = '?', s->pos = -1;
+		if (longopts) { /* parse long options */
+			int k, n_exact = 0, n_partial = 0;
+			const ko_longopt_t *o = 0, *o_exact = 0, *o_partial = 0;
+			for (j = 2; argv[s->i][j] != '\0' && argv[s->i][j] != '='; ++j) {} /* find the end of the option name */
+			for (k = 0; longopts[k].name != 0; ++k)
+				if (strncmp(&argv[s->i][2], longopts[k].name, j - 2) == 0) {
+					if (longopts[k].name[j - 2] == 0) ++n_exact, o_exact = &longopts[k];
+					else ++n_partial, o_partial = &longopts[k];
+				}
+			if (n_exact > 1 || (n_exact == 0 && n_partial > 1)) return '?';
+			o = n_exact == 1? o_exact : n_partial == 1? o_partial : 0;
+			if (o) {
+				s->opt = opt = o->val, s->longidx = o - longopts;
+				if (argv[s->i][j] == '=') s->arg = &argv[s->i][j + 1];
+				if (o->has_arg == 1 && argv[s->i][j] == '\0') {
+					if (s->i < argc - 1) s->arg = argv[++s->i];
+					else opt = ':'; /* missing option argument */
+				}
+			}
+		}
+	} else { /* a short option */
+		const char *p;
+		if (s->pos == 0) s->pos = 1;
+		opt = s->opt = argv[s->i][s->pos++];
+		p = strchr((char*)ostr, opt);
+		if (p == 0) {
+			opt = '?'; /* unknown option */
+		} else if (p[1] == ':') {
+			if (argv[s->i][s->pos] == 0) {
+				if (s->i < argc - 1) s->arg = argv[++s->i];
+				else opt = ':'; /* missing option argument */
+			} else s->arg = &argv[s->i][s->pos];
+			s->pos = -1;
+		}
+	}
+	if (s->pos < 0 || argv[s->i][s->pos] == 0) {
+		++s->i, s->pos = 0;
+		if (s->n_args > 0) /* permute */
+			for (j = i0; j < s->i; ++j)
+				ketopt_permute(argv, j, s->n_args);
+	}
+	s->ind = s->i - s->n_args;
+	return opt;
+}
+
+#endif
\ No newline at end of file
diff --git a/modules/audio_processing/around_runner/mocks.h b/modules/audio_processing/around_runner/mocks.h
new file mode 100644
index 0000000000000000000000000000000000000000..b35fa7de051a7f843b216d4322f5cafd2c68067d
--- /dev/null
+++ b/modules/audio_processing/around_runner/mocks.h
@@ -0,0 +1,64 @@
+#ifndef MODULES_AUDIO_PROCESSING_MOCKS_H_
+#define MODULES_AUDIO_PROCESSING_MOCKS_H_
+
+#include <string>
+
+namespace webrtc {
+  namespace field_trial {
+    bool IsEnabled(const char *)
+    {
+        return false;
+    }
+
+    std::string FindFullName(const std::string &)
+    {
+        return "";
+    }
+  }
+
+  namespace metrics {
+    class Histogram;
+
+    // Get histogram for counters.
+    Histogram* HistogramFactoryGetCounts(const std::string& ,
+                                        int ,
+                                        int ,
+                                        int )
+    {
+      return nullptr;
+    }
+
+    // Get histogram for counters with linear bucket spacing.
+    Histogram* HistogramFactoryGetCountsLinear(const std::string& ,
+                                              int ,
+                                              int ,
+                                              int )
+    {
+      return nullptr;
+    }
+
+    // Get histogram for enumerators.
+    // |boundary| should be above the max enumerator sample.
+    Histogram* HistogramFactoryGetEnumeration(const std::string& ,
+                                              int )
+    {
+      return nullptr;
+    }
+
+    // Get sparse histogram for enumerators.
+    // |boundary| should be above the max enumerator sample.
+    Histogram* SparseHistogramFactoryGetEnumeration(const std::string& ,
+                                                    int )
+    {
+      return nullptr;
+    }
+
+    // Function for adding a |sample| to a histogram.
+    void HistogramAdd(Histogram* , int )
+    {
+      return;
+    }
+  }
+}
+
+#endif
\ No newline at end of file
